File: go/resistor-color-duo/resistor_color_duo.go
// Package resistorcolorduo implements the Resistor Color Duo exercise.
package resistorcolorduo

const ignoreBandsAfterSecond = true

// Value returns the resistance value of a resistor with given colors.
func Value(colors []string) int {
	length := len(colors)
	if length < 2 {
		return -1
	}

	if ignoreBandsAfterSecond {
		length = 2
	}

	resistances := make([]int, length)

decode:
	for resistance, color := range allColors() {
		for i, band := range colors {
			if ignoreBandsAfterSecond && i > 1 {
				continue decode
			}
			if band == color {
				resistances[i] = resistance
			}
		}
	}

	totalResistance := 0
	for i, resistance := range resistances {
		totalResistance += resistance * ipow(10, length-1-i)
	}

	return totalResistance
}

func allColors() []string {
	return []string{"black", "brown", "red", "orange", "yellow", "green", "blue", "violet", "grey", "white"}
}

func ipow(base, exp int) int {
	if exp < 0 {
		panic("Negative exponent")
	}

	result := 1
	for {
		if exp%2 == 1 {
			result *= base
		}
		exp /= 2
		if exp == 0 {
			break
		}
		base *= base
	}

	return result
}

File: prolog/resistor-color-duo/resistor_color_duo.pl
:- module(resistor_color_duo, [value/2]).  /*

\PrologDialect{swiprolog}

\Predicate color_code/2(+Color:string, -Codeint).

True if \verb|Code| is the encoding of \verb|Color|.

\PL*/
%! color_code(+Color:string, -Code:int) is semidet.
%
% True if =Code= is the encoding of =Color=.
color_code("black",  0).
color_code("brown",  1).
color_code("red",    2).
color_code("orange", 3).
color_code("yellow", 4).
color_code("green",  5).
color_code("blue",   6).
color_code("violet", 7).
color_code("grey",   8).
color_code("white",  9).
/*PL

\Predicate value/2(+Bands:list(string), -Value:int).

The encoded \verb|Value| of the first two \verb|Bands| of a resistor.

\PL*/
%! value(+Bands:list(string), -Value:int) is semidet.
%
% The encoded =Value= of the first two =Bands= of a resistor.
value([Color1, Color2], Value) :-
    color_code(Color1, Code1),
    color_code(Color2, Code2),
    Value is 10 * Code1 + Code2,
    !.
value([Color1, Color2|_], Value) :-
    value([Color1, Color2], Value).
/*PL

\EndProlog*/
